base/mem_trie.hpp

```diff
@@ -177,49 +288,129 @@ class MemTrie
   template <typename ToDo>
   void ForEachInNode(String const & prefix, ToDo && toDo) const
   {
-    if (auto const * root = MoveTo(prefix))
-      ForEachInNode(*root, prefix, std::forward<ToDo>(toDo));
+    MoveTo(prefix, true /* fullMatch */,
+           [&](Node const & node, Edge const & /* edge */, size_t /* offset */) {
+             node.m_values.ForEach(std::forward<ToDo>(toDo));
+           });
   }
 
   // Calls |toDo| for each key-value pair in a subtree that is
   // reachable by |prefix| from the trie root. Does nothing if such
   // subtree does not exist.
   template <typename ToDo>
-  void ForEachInSubtree(String prefix, ToDo && toDo) const
+  void ForEachInSubtree(String const & prefix, ToDo && toDo) const
+  {
+    MoveTo(prefix, false /* fullMatch */, [&](Node const & node, Edge const & edge, size_t offset) {
+      String p = prefix;
+      for (; offset < edge.Size(); ++offset)
+        p.push_back(edge[offset]);
+      ForEachInSubtree(node, p, std::forward<ToDo>(toDo));
+    });
+  }
+
+  bool HasKey(String const & key) const
   {
-    if (auto const * root = MoveTo(prefix))
-      ForEachInSubtree(*root, prefix, std::forward<ToDo>(toDo));
+    bool exists = false;
+    MoveTo(key, true /* fullMatch */,
+           [&](Node const & node, Edge const & /* edge */, size_t /* offset */) {
+             exists = !node.m_values.Empty();
+           });
+    return exists;
   }
 
-  void Clear()
+  bool HasPrefix(String const & prefix) const
   {
-    m_root.Clear();
-    m_numNodes = 1;
+    bool exists = false;
+    MoveTo(prefix, false /* fullMatch */, [&](Node const & node, Edge const & /* edge */,
+                                              size_t /* offset */) { exists = !node.Empty(); });
+    return exists;
   }
 
-  size_t GetNumNodes() const { return m_numNodes; }
+  void Clear() { m_root.Clear(); }
+
+  size_t GetNumNodes() const { return m_root.GetNumNodes(); }
+
   Iterator GetRootIterator() const { return Iterator(m_root); }
+
   Node const & GetRoot() const { return m_root; }
 
 private:
+  class Edge
+  {
+  public:
+    Edge() = default;
+
+    template <typename It>
+    Edge(It begin, It end)
+    {
+      Assign(begin, end);
+    }
+
+    template <typename It>
+    void Assign(It begin, It end)
+    {
+      m_label.assign(begin, end);
+      std::reverse(m_label.begin(), m_label.end());
+    }
+
+    Edge Drop(size_t n)
```

Comment:
s/Drop/DropFirst/

Reply:
No, thanks. This is a common notation.

Reply:
I'm not sure it's that common.
Even in functional languages (where it makes sense, since dropping a suffix of an infinite list is meaningless), e.g. Erlang avoids using plain "drop" and has "droplast" (albeit with a different semantics): http://erlang.org/doc/man/lists.html#droplast-1

Up to you, but at least add a comment.

Reply:
Let's rename to DropFirst.